gtsam/navigation/tests/testAHRSFactor.cpp

```diff
@@ -174,17 +174,22 @@ TEST(AHRSFactor, Error) {
 
   // Expected Jacobians
   Matrix H1e = numericalDerivative11<Vector3, Rot3>(
-      boost::bind(&callEvaluateError, factor, _1, x2, bias), x1);
+      boost::bind(&callEvaluateError, factor, boost::placeholders::_1, x2,
```

Comment:
Since you're using `boost::placeholders` in multiple places, you should use `using namespace boost::placeholders`. This should also reduce the diff size.

Reply:
`using namespace ...` should be used as little as possible as it clutters the scope. Even then it's recommend to never use the `using` directive (See: https://google.github.io/styleguide/cppguide.html#Namespaces).

I deemed that `numericalDerivative.h` had a large number of uses to justify it, whereas for the other files I did not believe it was necessary.

In fact, I would rather spend the time to change `numericalDerivative.h` to not have `using namespace` than the other way around. Let me know what you still think, I'll add `using namespace` if you want.

Reply:
What is your definition of large number of uses? For us, it is generally 4 or more.

In that case, you should use `using boost:: placeholder::_1` etc since it avoids the issues of namespace pollution and other negative aspects. This is an acceptable pattern.

Reply:
I still think with `using boost::placeholder::_1` at the global scope you are polluting with `_1`. With `using` I believe it is recommended to use it locally in methods and not at the file level. At that point it should reduce the diff size a good bit.

I'll go ahead and do that and we can see if the diff size in appropriate.
(Sorry I'm being so adamant on not using `using` ;))

Reply:
It's not polluting the namespace since we're using that specific variable and only accessing that one variable.
It would be namespace pollution if other, unused, variables were being included into the namespace.

Reply:
Ah maybe pollution was the wrong word to use there. I agree with what you are describing, but there is the case, where if we `using boost::placeholder::_1` say in a header file, than `_1` becomes associated with `boost::placeholder::_1` and you can't use a local `_1`.

Reply:
I've updated the PR, where for tests I am adding `using namespace boost::placeholders;` as those are the end of the compilation pipeline. But for the other files, I've kept them as is. In fact, I doubled down on my stance for `numericalDerivative.h` and added the `using namespace boost::placeholders;` in the local scopes of the methods.

Reply:
Will we ever use a local `_1` in any context? We prefer to use descriptive names so I don't see that being a problem. I still don't agree with your use of the full namespace everywhere, even in headers, the reasons being that the placeholders are not part of the public API (especially if you do `using boost::placeholders::_1;` inside the `gtsam` namespace) and the use of `using boost::placeholders::_1;` is a common pattern we have used in other code bases in the lab. There is little reason to make the code more verbose for the sake of being pedantic.

Also, you have included the header in test files which don't use the placeholder, and not included in test files that do. What's going on there?

Reply:
> especially if you do `using boost::placeholders::_1` inside the gtsam namespace

I think this is fine, if done in cpp files, but should still not be done in header files. Here is a source on why: Page 109 [here](http://library.bagrintsev.me/CPP/Sutter.C%2B%2B%20Coding%20Standards.2005.pdf)

> Also, you have included the header in test files which don't use the placeholder, and not included in test files that do. What's going on there?

Can you point to these specifically? I thought I had added and removed the header appropriately, I guess I messed up for some of them. (Surprised the CI passes)

Reply:
Again, no reason to be pendantic about the standards if it makes the code unnecessarily verbose. We prefer readability over everything else. And as I mentioned above, the placeholders are not used as part of any public API.

Reply:
> the placeholders are not used as part of any public API

I'm a bit confused about this. What are you referring to as the public API?

Reply:
Are you referring to the header files users have access to? If so `gtsam/base/numericalDerivative.h` is a public header file.

Here is a minimal example, where the user would be frustrated.

exe.cpp
```
#include <gtsam/base/numericalDerivative.h>

#include <functional>
#include <iostream>
#include <typeinfo>

using namespace std::placeholders;
using namespace gtsam;

int main() { std::cout << typeid(_1).name() << std::endl; }
```

Having `using boost::placeholders::_1;` inside of the gtsam namespace in the header file of `numericalDerivative.h`.
Output:
```
/home/acxz/test-gtsam/exe.cpp:10:34: error: reference to ‘_1’ is ambiguous
   10 | int main() { std::cout << typeid(_1).name() << std::endl; }
      |                                  ^~
In file included from /usr/include/boost/bind/bind.hpp:2336,
                 from /usr/local/include/gtsam/base/numericalDerivative.h:22,
                 from /home/acxz/test-gtsam/exe.cpp:1:
/usr/include/boost/bind/placeholders.hpp:46:38: note: candidates are: ‘constexpr const boost::arg<1> boost::placeholders::_1’
   46 | BOOST_INLINE_CONSTEXPR boost::arg<1> _1;
      |                                      ^~
In file included from /usr/include/boost/function.hpp:23,
                 from /usr/local/include/gtsam/base/numericalDerivative.h:21,
                 from /home/acxz/test-gtsam/exe.cpp:1:
/usr/include/c++/11.1.0/functional:225:34: note:                 ‘const std::_Placeholder<1> std::placeholders::_1’
  225 |     extern const _Placeholder<1> _1;
      |                                  ^~
```

Having the `using` directive inside the scope of functions in the header file of `numericalDerivative.h`.
Output:
```
St12_PlaceholderILi1EE
```

As you can see the current change works as expected and returns the proper type, `std::placeholders::_1`. The first output gives us an ambiguous usage of `_1` and fails to compile.

Reply:
The user can then simply remove the first `using namespace boost::placeholders;` and things should work no problem. I don't really see what the problem is here.

Moreover, the `develop` branch currently uses `_1` and this has not been a problem all this time.

Reply:
> Moreover, the develop branch currently uses _1 and this has not been a problem all this time.

We actually get the same error with the current code.

> The user can then simply remove the first `using namespace boost::placeholders`;

The user's code does not have the line `using namespace boost::placeholders`; They are using `std::placeholders`.

Reply:
The error has not been reported, because the placeholder is not a part of the public API.

Well, you use std::placeholders and then mention boost::placeholders in the following comment. Either way I don't see how that makes a difference.

Reply:
The issue with your example is that you have a `using namespace gtsam` which directly contradicts your previous arguments. The user can simply use gtsam::X for whatever they need and issue resolved.

Reply:
> the placeholder is not a part of the public API.

I'm still a bit confused by this. By public API do you mean the header files that users have access to?

> Boost or std doesn't matter right? They can either use the boost one,

I'm actually not sure about that. There might be differences in std::bind and boost::bind, where it does matter, there might not be. (I personally feel like it doesn't.) The user will most likely err on the safer side if they are using `std::bind` and would want to use `std::placeholders::`.

> fully qualify the std one
> The user can simply use gtsam::X for whatever they need and issue resolved.

The user should not be at fault for having to use `using namespace` in `.cpp` files, especially code that is at the end of the compilation units, such as this example or tests. It is up to the libraries included to ensure included namespaces do not cause pollution or collisions. It is one thing if the user is also using `using boost::placeholders` along with `using std::placeholders` that's on them to fix the collision. But it is another thing if the collision error is coming from a library that they do not have control over. Sure they can fix the problem on their end, but they should not have experienced the problem in the first place if the library is created properly.

Reply:
> By public API do you mean the header files that users have access to?

By public API I mean that they will never need to use `_1` when calling a function in GTSAM. So they never have to deal with it wrt GTSAM code. The public API is any function that can be called by the user.

> The user should not be at fault for having to use using namespace in .cpp files, especially code that is at the end of the compilation units, such as this example or tests. It is up to the libraries included to ensure included namespaces do not cause pollution or collisions. It is one thing if the user is also using using boost::placeholders along with using std::placeholders that's on them to fix the collision. But it is another thing if the collision error is coming from a library that they do not have control over. Sure they can fix the problem on their end, but they should not have experienced the problem in the first place if the library is created properly.

Then the user can just be more specific and declare `using std::placeholders::_1;`. Again, I don't see what the problem is. We are not restricting them in any way.

Reply:
> they will never need to use _1 when calling a function in GTSAM

Can they not use the functions in numericalDerivatives which you can pass placeholder arguments?

> Then the user can just be more specific and declare `using std::placeholders::_1;` This is the same as `using namespace std::placeholders;`. The `_1` is imported in both cases, leading to the same collision problem.

It seems that we are at a standstill. I suggest we wait for another opinion on the matter, before taking this conversation further.